Applications/Spire/Source/Ui/TextAreaBox.cpp

```diff
@@ -0,0 +1,630 @@
+    const std::function<void(const QTextBlock&)> format) {
+  auto cursor_pos = m_text_edit->textCursor().position();
+  for(auto i = 0; i < m_text_edit->document()->blockCount(); ++i) {
+    auto block = m_text_edit->document()->findBlockByNumber(i);
+    if(block.isValid()) {
+      format(block);
+    }
+  }
+  auto cursor = m_text_edit->textCursor();
+  cursor.setPosition(cursor_pos);
+  m_text_edit->setTextCursor(cursor);
+}
+
+void TextAreaBox::commit_placeholder_style() {
+  m_placeholder->set_alignment(m_placeholder_styles.m_alignment);
+  auto font = m_placeholder_styles.m_font;
+  if(m_placeholder_styles.m_size) {
+    font.setPixelSize(*m_placeholder_styles.m_size);
+  }
+  m_placeholder->setFont(font);
+  m_placeholder->set_text_color(m_placeholder_styles.m_color);
+  update_placeholder_text();
+}
+
+void TextAreaBox::commit_style() {
+  auto stylesheet = QString(
+    R"(QTextEdit {
+      background: transparent;
+      border-width: 0px;)");
+  m_text_edit_styles.m_styles.write(stylesheet);
+  if(m_text_edit_styles.m_alignment!= m_text_edit->alignment()) {
+    update_text_alignment();
+  }
+  auto font = m_text_edit_styles.m_font;
+  if(m_text_edit_styles.m_size) {
+    font.setPixelSize(*m_text_edit_styles.m_size);
+  }
+  m_text_edit->setFont(font);
+  if(m_text_edit_styles.m_line_height &&
+      (font.pixelSize() * *m_text_edit_styles.m_line_height) !=
+      m_computed_line_height) {
+    update_line_height();
+  }
+  if(stylesheet != m_text_edit->styleSheet()) {
+    m_text_edit->setStyleSheet(stylesheet);
+    m_text_edit->style()->unpolish(this);
+    m_text_edit->style()->polish(this);
+  }
+}
+
+bool TextAreaBox::is_placeholder_shown() const {
+  return !is_read_only() && m_model->get_current().isEmpty() &&
+    !m_placeholder_text.isEmpty();
+}
+
+QSize TextAreaBox::get_border_size() const {
+  auto borders = m_text_edit_styles.m_border_sizes;
+  return {borders.left() + borders.right(), borders.top() + borders.bottom()};
+}
+
+QSize TextAreaBox::get_padding_size() const {
+  auto padding = m_text_edit_styles.m_padding;
+  return {padding.left() + padding.right(), padding.top() + padding.bottom()};
+}
+
+void TextAreaBox::update_display_text() {
+  if(is_read_only()) {
+    if(m_text_edit->toPlainText() != m_model->get_current()) {
+      m_text_edit->blockSignals(true);
+      m_text_edit->setText(m_model->get_current());
+      m_text_edit->blockSignals(false);
+    }
+    auto line_count = std::floor((height() - get_padding_size().height()) /
+      static_cast<double>(m_computed_line_height));
+    auto is_elided = false;
+    if(line_count > 0) {
+      auto lines = QStringList();
+      for(auto i = 0; i < m_text_edit->document()->blockCount() && !is_elided;
+          ++i) {
+        auto block = m_text_edit->document()->findBlockByNumber(i);
+        if(block.isValid()) {
+          auto block_text = block.text();
+          for(int i = 0; i != block.layout()->lineCount(); ++i) {
```

Comment:
The inner loop and the outer loop have the same iterator name "i", even though it is allowed by the standard, I don't think it's a good practice.